### PR TITLE
Add proposal for kubelet TLS bootstrap

### PR DESCRIPTION
**Pull request by gtank**
A proposal based on the discussion of issue #18112, to implement a process by which kubelets can obtain TLS certificates in a streamlined manner.

### CODE REVIEW

**Comment by smarterclayton** [comment]
@liggitt and @deads2k, parallels lots of discussions on cluster CSR

**Comment by smarterclayton** [comment]
What defense does this API need against abuse?

**Comment by gtank** [comment]
The abuse concerns I've considered are 1) DoS by making excessive requests and 2) unauthorized certificate issuance. We mitigate the first by restricting the endpoint itself with token auth (on the assumption that nodes you have provisioned are unlikely to DoS you) and the second by requiring manual intervention to approve CSRs.

In the future, we'd like to implement more elaborate approval policies (e.g. involving ACME handshakes or TPM attestation) on top of this mechanism.

File: docs/proposals/kubelet-tls-bootstrap.md
Comment (mikedanese) [inline docs/proposals/kubelet-tls-bootstrap.md]: @erictune what would be a good API group to add this too? Will we have an authentication api group eventually to house user and group objects? That seems like a reasonable place to put these types.

**Comment by mikedanese** [comment]
I assigned to you to do an initial review since it's auth related.

cc @roberthbailey @bgrant0607 @davidopp @jbeda

File: docs/proposals/kubelet-tls-bootstrap.md
Comment (liggitt) [inline docs/proposals/kubelet-tls-bootstrap.md]: This is more pki/ca-related than auth

**Comment by jbeda** [comment]
This is a little left field, but have we considered perhaps using/adapting the ACME protocol used by Let's Encrypt?

https://tools.ietf.org/html/draft-ietf-acme-acme-01

It is very similar to the flow in the doc but with some important differences:
* There idea of a user that has control over an identity is different from the requests for certificates.  This opens the door for certificate rotation and requests for different types of (perhaps down-scoped) certs.
* There are an extensible set of "challenges" that can be used to verify the client.  This opens the door for more sophisticated identity proof systems in the future.
* This opens the door for the CA to be run separately from the master with few/no changes to the kubelet/client.  This may be an important security consideration.

A lot of the block and tackle for LetsEncrypt is written in go already -- https://github.com/letsencrypt/boulder.  While we probably won't want to use this stuff verbatim, it may be possible to break out parts of the implementation and share (etcd raft-like).

My motivation: I'd love to see an extensible identity system that is "dialtone" in clusters not just for the cluster system itself but for applications.  Basically, we'll have the problem of "microservice A calls microservice B.  How does B know it is actually A that is calling".  This is obviously out of scope for k8s but it would be great to be compatible and share code.

File: docs/proposals/kubelet-tls-bootstrap.md
Comment (lavalamp) [inline docs/proposals/kubelet-tls-bootstrap.md]: I'd recommend adding a /approve subresource for this-- we don't ordinarily do a bunch of computation when you write to an object.

A /reject subresource isn't necessary for this, since in that case the apiserver is just filling in fields, as in an ordinary update.

File: docs/proposals/kubelet-tls-bootstrap.md
Comment (lavalamp) [inline docs/proposals/kubelet-tls-bootstrap.md]: `kubectl get csrs`-- signing requests are not certificates.

Comment (lavalamp) [inline docs/proposals/kubelet-tls-bootstrap.md]: What if you want to revoke a cert?

**Comment by smarterclayton** [comment]
I think it's in scope for the ecosystem to build up something capable
of allowing that. Agree it may not be Kube core, but it's pretty
fundamental.

**Comment by jbeda** [comment]
@smarterclayton -- Wrt app->app authn, I'm thinking that something as part of the CNCF might make sense.  I'm going to write a design doc over the next couple of weeks and share it around.  Stay tuned.

File: docs/proposals/kubelet-tls-bootstrap.md
Comment (smarterclayton) [inline docs/proposals/kubelet-tls-bootstrap.md]: I don't want to scope creep this, but I would like a section that describes how other usecases for CSR on the cluster overlaps, and make sure this is not so narrowly scoped that it can't be useful elsewhere.

File: docs/proposals/kubelet-tls-bootstrap.md
Comment (smarterclayton) [inline docs/proposals/kubelet-tls-bootstrap.md]: @deads2k can you describe service auto signing here (as prototyped) so we have some context

**Comment by smarterclayton** [comment]
Sgtm - @deads2k is off prototyping auto serving cert generation for
services and we have a few other investigations that would benefit
from a broader discussion and working group.

File: docs/proposals/kubelet-tls-bootstrap.md
Comment (deads2k) [inline docs/proposals/kubelet-tls-bootstrap.md]: Why are these exploded out?  They can't be trusted as authoritative since only the signed copy is immutable.  Are they derived and clients never submit then?  If so, why not put them on status?

File: docs/proposals/kubelet-tls-bootstrap.md
Comment (deads2k) [inline docs/proposals/kubelet-tls-bootstrap.md]: This seems like a spot where people are likely to put confidential information, things like: "this is proof that I'm allowed to make this kind of a request".  If that's the case, then it seems like this should include a secret reference.

File: docs/proposals/kubelet-tls-bootstrap.md
Comment (deads2k) [inline docs/proposals/kubelet-tls-bootstrap.md]: I think I'd expect the API server to keep track of who created the `CertificateSigningRequest` in a field that normal caller can't touch so that information could be used in making an approval decision.

File: docs/proposals/kubelet-tls-bootstrap.md
Comment (deads2k) [inline docs/proposals/kubelet-tls-bootstrap.md]: Is the API server doing the signing or is that being done by a controller which then does a straight update?  The latter seems more "natural", but it also suggests that `ConditionStatus` isn't rich enough to describe the states which appear to be:
 1. approval-pending
 1. approved
 1. denied
 1. signed

File: docs/proposals/kubelet-tls-bootstrap.md
Comment (deads2k) [inline docs/proposals/kubelet-tls-bootstrap.md]: If we have a subresource that we expect administrators to hit, why not extend our existing infrastructure to easily operate against subresources?  Seems like this would be an update to `csr/approve`.

**Comment by mikedanese** [comment]
It looks like let's encrypt is using https://github.com/cloudflare/cfssl as a certificate authority and for a bunch of other things. Written in go, so might be of interest.

**Comment by deads2k** [comment]
> @deads2k can you describe service auto signing here (as prototyped) so we have some context

One problem we've had is being able to create and sign serving certificates that can be used by pods backing a particular service.  In many simple cases an RC backs one service and would like to serve securely, but it needs a cert signed for the service DNS name for each pod.  That can be done with a secret, but building the right secrets in the right format is hard and users don't care about the particular key/cert to use.

Given an annotation on a service, a key/cert pair can be create, but signing it in a reasonable way is a little more difficult.  I think that this mechanism could serve, given a sufficiently rich "deny" plug-point.

File: docs/proposals/kubelet-tls-bootstrap.md
Comment (deads2k) [inline docs/proposals/kubelet-tls-bootstrap.md]: Can you elaborate on what would go into "a basic config file"?

**Comment by gtank** [comment]
@jbeda In my mind, this is an adaptation of the ACME protocol. Originally I wanted to write a challenge based on ssh to auto-approve node certificates, but commentary on my initial document suggested scoping it down to a more flexible Kubernetes-specific process. I still think of "wait for manual approval" as a degenerate case of ACME challenge, and could phrase it that way if it would be better.

@mikedanese cfssl is a good project. I've done some work with it before and would absolutely recommend it for more complicated CA features in the future. But if we don't want to carry the dependency yet, the Go stdlib gives us everything we need for basic CSR approval.

File: docs/proposals/kubelet-tls-bootstrap.md
Comment (gtank) [inline docs/proposals/kubelet-tls-bootstrap.md]: sgtm. I'm not terribly familiar with the API conventions yet and this still struck me as a rough edge.

File: docs/proposals/kubelet-tls-bootstrap.md
Comment (gtank) [inline docs/proposals/kubelet-tls-bootstrap.md]: Can you point me to more info about your use cases? This is certainly general enough for other purposes, but node init is the main one we have.

**Comment by erictune** [comment]
Joe's suggestion of using ACME makes sense to me.   Presumably a lot of work went into getting bits of the standard right.

Having a non-kubernetes-style endpoint served off of the apiserver is fine.  We already do this in a few places.

File: docs/proposals/kubelet-tls-bootstrap.md
Comment (gtank) [inline docs/proposals/kubelet-tls-bootstrap.md]: They're intended to be derived for convenience of filtering / display. If we're going to do anything to them, it should probably just be to remove them. You correctly point out that they can't be trusted. But does an API object made of opaque bytes have precedent?

**Comment by jbeda** [comment]
@gtank I don't want to hijack your issue if we've already had the ACME discussion.  As for some of the points:
* Agree that manual approval fits in with ACME and is a degenerate case
* I'm imagining the account key holder would be the machine.  No reasons for those keys to ever leave the machine if we can help it.  Each machine would have a separate account.  No need for the recovery stuff or EULA or contact info.

File: docs/proposals/kubelet-tls-bootstrap.md
Comment (deads2k) [inline docs/proposals/kubelet-tls-bootstrap.md]: > They're intended to be derived for convenience of filtering / display. If we're going to do anything to them, it should probably just be to remove them. You correctly point out that they can't be trusted.

Maybe comments about not being user settable?  You can enforce that in a rest storage strategy.  `Spec` versus `Status` would be a @bgrant0607 of @lavalamp kind of question, but written like this the trustability of this field isn't clear.

File: docs/proposals/kubelet-tls-bootstrap.md
Comment (gtank) [inline docs/proposals/kubelet-tls-bootstrap.md]: This makes sense to me in the general case. For node init though, this is going to be a mechanism for bootstrapping identity. Can secrets still be created in a reasonable way using pre-shared token auth?

File: docs/proposals/kubelet-tls-bootstrap.md
Comment (lavalamp) [inline docs/proposals/kubelet-tls-bootstrap.md]: I suggest: `/apis/certificates/v1beta1/signingrequests/mycsr`

File: docs/proposals/kubelet-tls-bootstrap.md
Comment (lavalamp) [inline docs/proposals/kubelet-tls-bootstrap.md]: I suggest:
* include only things that the client can set in Spec (e.g., CertificateRequest)
* make it clear that this type's spec is immutable
* if you have a use case for these breakout helpers, put them in status.

**Comment by bgrant0607** [comment]
Ref #5754

File: docs/proposals/kubelet-tls-bootstrap.md
Comment (gtank) [inline docs/proposals/kubelet-tls-bootstrap.md]: Agree 100%. We can make the CSR bytes immutable after creation and leave everything else not user settable.

File: docs/proposals/kubelet-tls-bootstrap.md
Comment (gtank) [inline docs/proposals/kubelet-tls-bootstrap.md]: I have it under `extensions` for now. Will move it to a more appropriate location when we can figure out where that is.

File: docs/proposals/kubelet-tls-bootstrap.md
Comment (gtank) [inline docs/proposals/kubelet-tls-bootstrap.md]: @lavalamp It's hard to come up with a good revocation story here. We'd need to at least add the notion of a CRL to the apiserver and more kubectl commands for dealing with that. My intent is to make the (configurable) default a relatively short-lifetime certificate with a provision for automatic renewal, like an ACME proof of possession. But I think this is for after the basic feature is in place, and have listed it under "future work".

@deads2k Not sure what you mean by "existing infrastructure" in this case.

File: docs/proposals/kubelet-tls-bootstrap.md
Comment (gtank) [inline docs/proposals/kubelet-tls-bootstrap.md]: That sounds reasonable. Since this is also intended as a way to establish node identity, what information should we include? IP address?

Comment (liggitt) [inline docs/proposals/kubelet-tls-bootstrap.md]: I would not expect this to be part of identity. I would expect each node to be given a kubeconfig file with client credentials it could use to register, request serving certs, etc. That would mean the requesting username should be available to the rest handler impl.

File: docs/proposals/kubelet-tls-bootstrap.md
Comment (gtank) [inline docs/proposals/kubelet-tls-bootstrap.md]: After trying both options, I agree that the controller approach is better here. So it will be a straight update from the controller, but I think we should still stick with a simpler status. I think the right three are:

1. Submitted
2. Approved and available
3. Denied

What would "approved-but-not-yet-signed" mean to a consumer of this API?

File: docs/proposals/kubelet-tls-bootstrap.md
Comment (gtank) [inline docs/proposals/kubelet-tls-bootstrap.md]: As a first pass, information about what's an acceptable subject (CN is not "admin", or only *.cluster.domain.com) and how long the certificates should be valid.

File: docs/proposals/kubelet-tls-bootstrap.md
Comment (gtank) [inline docs/proposals/kubelet-tls-bootstrap.md]: My goal is to avoid having to provision a node with anything other than the location of its apiserver (or even a bootstrap server as in https://github.com/kubernetes/kubernetes/issues/23172). Given that, a node will be able to request a certificate with this API, then switch to TLS auth with meaningful subject information for further communication (including registration) with the apiserver. This allows you to write more robust ABAC policy and enforce clustering restrictions with arbitrary admission controllers. We're building one based on TPM attestation, but it could be anything.

In this scenario, I'm not sure where a username would come from. A field for any relevant caller ID sounds very useful though- for direct API access (kubectl/curl) I'd expect client credentials, for services in pods I'd expect a service account, etc. Would that do what you need?

File: docs/proposals/kubelet-tls-bootstrap.md
Comment (deads2k) [inline docs/proposals/kubelet-tls-bootstrap.md]: > @deads2k Not sure what you mean by "existing infrastructure" in this case.

You're talking about operating against a subresource.  It seems like what I really want is to `kubectl create csrs/request-name/approve` with some data.  Why wouldn't we start by allowing manipulation with our primitives before creating one off commands?

File: docs/proposals/kubelet-tls-bootstrap.md
Comment (deads2k) [inline docs/proposals/kubelet-tls-bootstrap.md]: > I have it under extensions for now. Will move it to a more appropriate location when we can figure out where that is.

No more under extensions please.  We can argue over the exact name in the pull or here, but I don't want any more "temporary" resources under extensions.

File: docs/proposals/kubelet-tls-bootstrap.md
Comment (deads2k) [inline docs/proposals/kubelet-tls-bootstrap.md]: > In this scenario, I'm not sure where a username would come from.

`user.Info` is a kube primitive available on the context of every authenticated request that includes the `username` of every requester.  You should respect that username.

We (openshift) would probably have a node present certificate credentials, since those shouldn't be compromised by communicating with an untrusted server and that would allow us secure this request endpoint to authenticated users only.

Were you envisioning this running exposed to unauthenticated requests that cannot be traced back to a requester?  Why would an admin ever approve?  He'd have no proof that he's signing for something legitimate.

File: docs/proposals/kubelet-tls-bootstrap.md
Comment (deads2k) [inline docs/proposals/kubelet-tls-bootstrap.md]: > I think the right three are:
> 
> Submitted
> Approved and available
> Denied

Those look fine.

File: docs/proposals/kubelet-tls-bootstrap.md
Comment (erictune) [inline docs/proposals/kubelet-tls-bootstrap.md]: If we are allowing IP SANs, don't we need a solid story on revokation, since IPs need to be recycled periodically?

File: docs/proposals/kubelet-tls-bootstrap.md
Comment (philips) [inline docs/proposals/kubelet-tls-bootstrap.md]: @smarterclayton Do you have other CSR use cases in mind?

Comment (gtank) [inline docs/proposals/kubelet-tls-bootstrap.md]: @erictune and I talked about this face-to-face and agreed that pod-level CSRs are out of scope for this proposal. They're necessary, but belong at a logically higher level than this (after clustering takes place).

File: docs/proposals/kubelet-tls-bootstrap.md
Comment (gtank) [inline docs/proposals/kubelet-tls-bootstrap.md]: I've been using cfssl in my prototyping work. They already define a policy file that supports all kinds of things we'll want in the future: https://github.com/cloudflare/cfssl/blob/master/config/config.go#L62

File: docs/proposals/kubelet-tls-bootstrap.md
Comment (gtank) [inline docs/proposals/kubelet-tls-bootstrap.md]: Maybe we don't need IP SANs at all? They're much more fungible than hostnames will be.

**Comment by gtank** [comment]
I've significantly revised this document and think I've addressed everyone's concerns. The most significant updates are 1) separating the immutable status information from the approval subresource and 2) ditching the dedicated kubectl commands for approval and rejection.

File: docs/proposals/kubelet-tls-bootstrap.md
Comment (mikedanese) [inline docs/proposals/kubelet-tls-bootstrap.md]: This should be a subobject is status. If you want to have a subresource that only modifies this field, that should still be fine. Also why not call this Approval?

File: docs/proposals/kubelet-tls-bootstrap.md
Comment (mikedanese) [inline docs/proposals/kubelet-tls-bootstrap.md]: what is a pkix.Name? we should import a representation of it into the api so we don't depend on non api packages.

File: docs/proposals/kubelet-tls-bootstrap.md
Comment (mikedanese) [inline docs/proposals/kubelet-tls-bootstrap.md]: States are not allowed in the API as they are not backwards compatible. These should be conditions.

File: docs/proposals/kubelet-tls-bootstrap.md
Comment (mikedanese) [inline docs/proposals/kubelet-tls-bootstrap.md]: This is redundant? Assuming a CSR exists in the API, it has been submitted?

File: docs/proposals/kubelet-tls-bootstrap.md
Comment (mikedanese) [inline docs/proposals/kubelet-tls-bootstrap.md]: json field names should be lower camelcase of struct field names

File: docs/proposals/kubelet-tls-bootstrap.md
Comment (mikedanese) [inline docs/proposals/kubelet-tls-bootstrap.md]: json field names should be lower camelcase of struct field names

File: docs/proposals/kubelet-tls-bootstrap.md
Comment (mikedanese) [inline docs/proposals/kubelet-tls-bootstrap.md]: This is basically a CertificateSigningRequestCondition

File: docs/proposals/kubelet-tls-bootstrap.md
Comment (mikedanese) [inline docs/proposals/kubelet-tls-bootstrap.md]: Maybe call this CertificateSigningRequestConditionType like https://github.com/kubernetes/kubernetes/blob/master/pkg/api/types.go#L1062-L1069

File: docs/proposals/kubelet-tls-bootstrap.md
Comment (gtank) [inline docs/proposals/kubelet-tls-bootstrap.md]: My idea was to separate the controller-derived information from the approval mechanism. In this arrangement, "Status" fields are immutable by users (and are generated from the signed CSR) and interaction takes place on "Approval" (which is a better name; I was thinking of `/approve` endpoints).

This structure also makes it easier to apply a separate registry strategy to the approval resource.

File: docs/proposals/kubelet-tls-bootstrap.md
Comment (gtank) [inline docs/proposals/kubelet-tls-bootstrap.md]: Basically a struct of strings representing the [Subject information from an X509 certificate](https://godoc.org/crypto/x509/pkix#Name). It's in the Go standard library, but if that still needs to be rolled in then where does it go?

File: docs/proposals/kubelet-tls-bootstrap.md
Comment (gtank) [inline docs/proposals/kubelet-tls-bootstrap.md]: Ah, I see the distinction. It could probably all be expressed as a condition- is there another example of action-gated-by-condition in the API?

File: docs/proposals/kubelet-tls-bootstrap.md
Comment (gtank) [inline docs/proposals/kubelet-tls-bootstrap.md]: Submitted is intended to express that the CSR has passed validation logic (which currently just checks it for a valid signature).

File: docs/proposals/kubelet-tls-bootstrap.md
Comment (mikedanese) [inline docs/proposals/kubelet-tls-bootstrap.md]: Scheduler doesn't schedule on nodes that don't have the ready condition. https://github.com/kubernetes/kubernetes/blob/master/pkg/api/types.go#L1754-L1765

File: docs/proposals/kubelet-tls-bootstrap.md
Comment (mikedanese) [inline docs/proposals/kubelet-tls-bootstrap.md]: Is validation logic check asynchronously from the create?

File: docs/proposals/kubelet-tls-bootstrap.md
Comment (mikedanese) [inline docs/proposals/kubelet-tls-bootstrap.md]: time.Duration and net.IP are in the standard library but we don't use them in the api. We have instead created types or use strings that we transform them to/from. I'd suggest creating a type for this in the certificates api group.

File: docs/proposals/kubelet-tls-bootstrap.md
Comment (mikedanese) [inline docs/proposals/kubelet-tls-bootstrap.md]: If you notice, most top level types only have ObjectMeta, TypeMeta, Spec, and sometimes Status. This is part of the api convention:

> All objects that represent a physical resource whose state may vary from the user's desired intent SHOULD have a "spec" and a "status". Objects whose state cannot vary from the user's desired intent MAY have only "spec", and MAY rename "spec" to a more appropriate name.

Desired intent is approved, so i think this cases falls squarely into the spec/status category where approval is part of Regardless you have to pick one or the other of these catagories, not both as you have now.

https://github.com/kubernetes/kubernetes/blob/master/docs/devel/api-conventions.md#spec-and-status

> This structure also makes it easier to apply a separate registry strategy to the approval resource.

You can still get access to .status.approval through an `/approve` subresource (i.e. a registry strategy can apply to a subfield, not just toplevel fields) which I think you implied with the `easier` modifier.

**Comment by gtank** [comment]
@mikedanese How about doing the conditions like this?

File: docs/proposals/kubelet-tls-bootstrap.md
Comment (mikedanese) [inline docs/proposals/kubelet-tls-bootstrap.md]: I think this filed should be moved to the body of status.

File: docs/proposals/kubelet-tls-bootstrap.md
Comment (mikedanese) [inline docs/proposals/kubelet-tls-bootstrap.md]: enums are namespaced in the api. e.g. this should be CertificateApproved and CertificateDenied

File: docs/proposals/kubelet-tls-bootstrap.md
Comment (mikedanese) [inline docs/proposals/kubelet-tls-bootstrap.md]: Many of these fileds don't really represent differences from desired state and should probably be moved into spec.

**Comment by mikedanese** [comment]
The api needs refining but this is going to be an alpha api for 1.3 so there will be significant api reviews later. LGTM'ing to unblock progress on an implementation.

File: docs/proposals/kubelet-tls-bootstrap.md
Comment (smarterclayton) [inline docs/proposals/kubelet-tls-bootstrap.md]: I didn't see a discussion (or missed it in a previous version) but can I watch the list and steal someone else's certificate?

Comment (liggitt) [inline docs/proposals/kubelet-tls-bootstrap.md]: Neither the csr nor the cert contains the private key of the node. A cert just holds the public key half, and isn't a secret (it's presented by the node in every TLS handshake)

Comment (smarterclayton) [inline docs/proposals/kubelet-tls-bootstrap.md]: Oh, right.

**Comment by ravilr** [comment]
@gtank Are there any plans to extend the CSR api proposed here, to support service->service mutual tls auth?  something like 
https://github.com/allingeek/pollendina#architecture

**Comment by smarterclayton** [comment]
We've had a fair number of discussions about how client certs /
serving certs might be generated.  I personally prefer approaches
where the infrastructure can sign for services to have certs injected,
but that does not directly solve the client problem.  I do like the
ideas endorsed in your link, of being able to request signing.

File: docs/proposals/kubelet-tls-bootstrap.md
Comment (munnerz) [reply docs/proposals/kubelet-tls-bootstrap.md]: Hi all - sorry to dig up an old thread!

I'm working on cert-manager, which is effectively 'kube-lego 2.0'. It aims to generalise obtaining certificates in Kubernetes, specifically for applications (which it looks like this comment thread touched on way back when).

We're looking at this moving into the kubernetes-incubator, and have a few open questions around existing work. I see the k8s certificates API, and we are finding ourselves effectively re-implementing it.

It seems like this document is the canonical design proposal for the Certificates API group itself, but I can't seem much information on the expected uses of the API.

As it stands, I can see a few issues with the existing certificates API that prevent us being able to adopt it:

* CSRs are non-namespaced - users are often scoped to only particular namespaces, but still need to make requests. There's no way for us to support this right now, and so users would need create/update/delete permission for all CSR resources, which breaks a lot of isolation models.

* CSRs directly embed signed certificates in their API status - this may not be too much of an issue, as the Certificate itself needn't be classed as 'secret'.

* Lack of ability to specify 'issuing source' - so if you look at the cert-manager proposal, we intend to support a variety of issuing sources. This includes a simple CA keypair (like the kube-controller-manager does right now), as well as others such as ACME, and in future potentially Vault. There is no way to 'choose' which issuer should issue a CSR as part of the certificates API, meaning users would have to go "all or nothing" with different issuers. In cert-manager, the `Certificate` resource (the equivalent of the `CertificateSigningRequest` resource) has an `issuer` field which allows us to disambiguate. We also allow extra configuration for the issuer to be specified (e.g. in the ACME case, which challenge mechanism should be used)

Comment (munnerz) [reply docs/proposals/kubelet-tls-bootstrap.md]: For reference, I have a draft proposal drawn up for cert-manager already: https://github.com/jetstack-experimental/cert-manager/pull/50

**Review by MrWoods024 — APPROVED** [review]

**Review by MrWoods024 — APPROVED** [review]

**Pull request merged**
